### PR TITLE
[CAZB-2223] Leeds taxi - include second week selection

### PR DESCRIPTION
**Pull request by pkalamarski**
## Motivation and Context

https://eaflood.atlassian.net/browse/CAZB-2223

## Description

Added second week selection view

## How Has This Been Tested?

Manually, rspec, cucumber

## Screenshots (if appropriate):
image
image
image

## Checklist:

- [x] It contains only changes required by issue (does not contain other PR)
- [x] Includes the link to an issue (if apply)
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: app/controllers/dates_controller.rb [hunk 1]
```diff
@@ -269,6 +277,67 @@ def confirm_select_weekly_period
...
 
+  ##
+  # Renders the list of dates to pick for respective week
+  #
+  # ==== Path for week one
+  #    GET /dates/select_weekly_date
+  #
+  # ==== Path for week two
+  #    GET /dates/select_second_weekly_date
+  #
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:288]: this is just helper method remove it and just describe something like this `validates weekly dates and assign variables used in views`

File: app/controllers/dates_controller.rb [hunk 2]
```diff
@@ -228,16 +237,18 @@ def select_weekly_date # rubocop:disable Metrics/AbcSize
   # * +la_name+ - lack of VRN redirects to {enter_details}[rdoc-ref:VehiclesController.enter_details]
   # * +charge+ - lack of VRN redirects to {enter_details}[rdoc-ref:VehiclesController.enter_details]
   #
-  def confirm_date_weekly
-    service = Dates::ValidateSelectedWeeklyDate.new(params: params, charge_start_date: @charge_start_date)
+  def confirm_date_weekly # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
+    is_second_week = !request.referer&.include?(select_weekly_date_dates_path)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:241]: Please add a new session manipulator and use it to set true or false if user already choose first week dates or second or even canceled to add a second week dates

File: app/controllers/charges_controller.rb [hunk 3]
```diff
@@ -72,15 +72,16 @@ def submit_local_authority
...
     @vrn = vrn
     @la_name = la_name
     @weekly_period = vehicle_details('weekly')
     @weekly_charge_today = vehicle_details('weekly_charge_today')
-    @dates = vehicle_details('dates')
+    @dates = vehicle_details('dates').sort { |a, b| a <=> b }
     @total_charge = vehicle_details('total_charge')
     @return_path = review_payment_return_path
     @chargeable_zones = vehicle_details('chargeable_zones')
+    @another_week_chosen = session[:taxi_second_week_start].present?
```
Comment (likeuwill) [inline app/controllers/charges_controller.rb:84]: just `@another_week_chosen = session[:taxi_second_week_start]`

File: app/controllers/dates_controller.rb [hunk 4]
```diff
@@ -269,6 +280,41 @@ def confirm_select_weekly_period
 
   private
 
+  ##
+  # Validates weekly dates and assigns variables used in views
+  def handle_select_weekly_date # rubocop:disable Metrics/AbcSize
+    is_second_week = request.path&.include?(select_second_weekly_date_dates_path)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:286]: same, get value from session

File: app/controllers/dates_controller.rb [hunk 5]
```diff
@@ -269,6 +280,41 @@ def confirm_select_weekly_period
...
+
+    if service.pay_week_starts_today? && vehicle_details('confirm_weekly_charge_today') != false
+      add_weekly_charge_today_to_session(service.today_date)
+      redirect_to select_weekly_period_dates_path
+    end
+  end
+
+  # Determinates redirect path after invalid date selected
+  def determinate_week_select_redirect_path
+    if request.referer&.include?(select_weekly_date_dates_path)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:311]: same, use value from session, u can add a helper method e.g. `second_week_selected?`

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -33,6 +33,8 @@ def determinate_payment_for_path(weekly_period, weekly_charge_today)
     if weekly_period && weekly_charge_today
       select_weekly_period_dates_path
     elsif weekly_period && !weekly_charge_today
+      session[:taxi_first_week_start] = nil
+      session[:taxi_second_week_start] = nil
```
Comment (likeuwill) [inline app/helpers/application_helper.rb:37]: i dont like naming, in future it's can be not only for taxi, can we call it e.g.
`first_week_start_date` and `second_week_start_date`

File: app/controllers/dates_controller.rb (hunk 2 above)
Comment (likeuwill) [reply app/controllers/dates_controller.rb:241]: please use variable `second_week_selected?` for manipulator and in controller

File: app/controllers/dates_controller.rb [hunk 7]
```diff
@@ -358,4 +404,8 @@ def add_confirm_charge_today_to_session
   def select_weekly_date_return_path
     vehicle_details('weekly_charge_today') ? select_weekly_period_dates_path : weekly_charge_dates_path
   end
+
+  def week_start_days
+    [session[:taxi_first_week_start], session[:taxi_second_week_start]].select(&:present?)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:409]:
```suggestion
    [session[:taxi_first_week_start], session[:taxi_second_week_start]].compact
```

File: app/services/session_manipulation/set_selected_week.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module SessionManipulation
+  ##
+  # Service used to set payment ID returned by the API.
+  #
+  # ==== Usage
+  #    SessionManipulation::SetPaymentId.call(session: session, payment_id: payment.payment_id)
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:8]: update rdoc

File: app/services/session_manipulation/set_selected_week.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    #
+    # ==== Attributes
+    # * +session+ - the user's session
+    #
+    def initialize(session:, week_selected: '')
+      @session = session
+      @week_selected = week_selected
+    end
+
+    # Adds the +payment_id+ to the session. Used by the class level method +.call+
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:24]: rdoc

File: app/controllers/dates_controller.rb [hunk 10]
```diff
@@ -229,15 +244,16 @@ def select_weekly_date # rubocop:disable Metrics/AbcSize
   # * +charge+ - lack of VRN redirects to {enter_details}[rdoc-ref:VehiclesController.enter_details]
   #
   def confirm_date_weekly
-    service = Dates::ValidateSelectedWeeklyDate.new(params: params, charge_start_date: @charge_start_date)
+    service = Dates::ValidateSelectedWeeklyDate.new(params: params,
+                                                    charge_start_date: @charge_start_date,
+                                                    session: session,
+                                                    is_second_week: second_week_being_selected?)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:250]: you already have all data in session, so you no need to send `is_second_week`

File: app/controllers/dates_controller.rb [hunk 11]
```diff
@@ -269,6 +285,39 @@ def confirm_select_weekly_period
 
   private
 
+  ##
+  # Validates weekly dates and assigns variables used in views
+  def handle_select_weekly_date
+    service = Dates::Weekly.new(vrn: vrn, zone_id: la_id, charge_start_date: @charge_start_date,
+                                is_second_week: second_week_being_selected?, week_start_days: week_start_days)
+
+    check_if_pay_week_starts_today(service, second_week_being_selected?)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:294]:
```
check_if_pay_week_starts_today(service) unless second_week_being_selected
```

File: app/controllers/dates_controller.rb [hunk 12]
```diff
@@ -358,4 +408,16 @@ def add_confirm_charge_today_to_session
...
+    [session[:first_week_start_date], session[:second_week_start_date]].compact
+  end
+
+  def reset_week_selection
+    SessionManipulation::SetSelectedWeek.call(session: session)
+  end
+
+  def second_week_being_selected?
+    session[:first_week_dates_selected]
+  end
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:422]: second or first ? :)

File: app/controllers/dates_controller.rb [hunk 13]
```diff
@@ -340,6 +389,7 @@ def determinate_next_weekly_page
     add_confirm_charge_today_to_session
     if params[:confirm_weekly_charge_today] == 'true'
       SessionManipulation::CalculateTotalCharge.call(session: session, weekly: true)
+      SessionManipulation::SetSelectedWeek.call(session: session, week_selected: 'first')
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:392]:
```suggestion
      SessionManipulation::SetSelectedWeek.call(session: session, second_week_selected: false)
```

File: app/views/dates/select_weekly_date/_content.html.haml [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+      - if @all_paid
+        .govuk-inset-text
+          If the weekly charge you want to pay for is not available, 
+          it is because you have paid for a date in those 7 consecutive days.
+
+      - if @d_day_notice
+        .govuk-inset-text
+          The Clean Air Zone charge came into operation on #{parse_single_date(@charge_start_date)}.
+
+      = form_tag confirm_date_weekly_dates_path(week: !second_week ? 'first' : 'second'), method: :post do
```
Comment (likeuwill) [inline app/views/dates/select_weekly_date/_content.html.haml:26]: i think you don't need to send which week was selected in params you already have it in session

File: app/services/session_manipulation/set_selected_week.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    def call
+      if @week_selected == 'first'
+        session[:first_week_dates_selected] = true
+      elsif @week_selected == 'second'
+        session[:second_week_dates_selected] = true
+      else
+        session[:first_week_dates_selected] = false
+        session[:second_week_dates_selected] = false
+      end
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:33]: Logic is too complicated, let me know if I'm wrong but I think you need only `second_week_selected?` variable in session.
So if first week selected you set to false, if second to true, and then you can check in services `nil, false or true`

File: app/controllers/dates_controller.rb (hunk 12 above)
Comment (pkalamarski) [reply app/controllers/dates_controller.rb:422]: I need this variable to tell `Dates for second week are CURRENTLY being chosen` - that means first week should be as **already** selected, but second shouldn't be as selected yet. It causes confusion for myself as well, so I think there could be a better naming

File: app/views/dates/select_weekly_date/_content.html.haml (hunk 14 above)
Comment (pkalamarski) [reply app/views/dates/select_weekly_date/_content.html.haml:26]: The session value is set based on this parameter

File: app/services/session_manipulation/set_selected_week.rb [hunk 16]
```diff
@@ -25,6 +25,14 @@ def initialize(session:, second_week_selected: false)
     # Adds the +second_week_dates_selected+ to the session. Used by the class level method +.call+
     def call
       session[:second_week_dates_selected] = @second_week_selected
+
+      reset_second_week_date unless @second_week_selected
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:29]:
```suggestion
      session[:second_week_start_date] = nil if @second_week_selected == false
```

File: app/views/charges/review_payment.html.haml [hunk 17]
```diff
@@ -52,3 +52,6 @@
                 = parse_charge(@total_charge)
               %td.govuk-table__cell
         = submit_tag 'Continue', class: 'govuk-button', 'data-module': 'govuk-button'
+
+        - unless @another_week_chosen
+          %p= link_to 'Add another week', select_second_weekly_date_dates_path, id: 'pay_instead'
```
Comment (likeuwill) [inline app/views/charges/review_payment.html.haml:57]: `id: 'add-another-week-link`

File: app/controllers/charges_controller.rb [hunk 18]
```diff
@@ -77,10 +77,11 @@ def review_payment
     @la_name = la_name
     @weekly_period = vehicle_details('weekly')
     @weekly_charge_today = vehicle_details('weekly_charge_today')
-    @dates = vehicle_details('dates')
+    @dates = vehicle_details('dates').sort
     @total_charge = vehicle_details('total_charge')
     @return_path = review_payment_return_path
     @chargeable_zones = vehicle_details('chargeable_zones')
+    @another_week_chosen = session[:second_week_start_date]
```
Comment (likeuwill) [inline app/controllers/charges_controller.rb:84]: Maybe better to call it `@another_week_dates_chosen`

File: app/controllers/dates_controller.rb [hunk 19]
```diff
@@ -229,15 +243,15 @@ def select_weekly_date # rubocop:disable Metrics/AbcSize
   # * +charge+ - lack of VRN redirects to {enter_details}[rdoc-ref:VehiclesController.enter_details]
   #
   def confirm_date_weekly
-    service = Dates::ValidateSelectedWeeklyDate.new(params: params, charge_start_date: @charge_start_date)
+    service = Dates::ValidateSelectedWeeklyDate.new(params: params,
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:246]: `params` we don't need it anymore

File: app/services/dates/validate_selected_weekly_date.rb [hunk 20]
```diff
@@ -62,7 +64,17 @@ def date_chargeable?
...
     def valid?
-      @start_date && date_in_range? && date_chargeable?
+      @start_date && date_in_range? && date_chargeable? && !already_selected?
+    end
+
+    ##
+    # Checks if date was already selected in previous week selection
+    # Returns boolean
+    def already_selected?
+      return unless @session[:second_week_dates_selected]
```
Comment (likeuwill) [inline app/services/dates/validate_selected_weekly_date.rb:74]:
```suggestion
      return if @session[:second_week_dates_selected].nil?
```

File: app/controllers/dates_controller.rb (hunk 19 above)
Comment (pkalamarski) [reply app/controllers/dates_controller.rb:246]: params are used in `app/services/dates/validate_selected_weekly_date.rb` (line 31) to get date from input

File: app/services/dates/weekly.rb [hunk 21]
```diff
@@ -10,17 +10,24 @@ module Dates
...
     # Build the list of dates and return them, e.g.
     # [{value: "2019-10-11", name: "Friday 11 October 2019", today: false},...]
     def chargeable_dates
-      @chargeable_dates ||= (calculated_start_date..(today + 6.days)).map { |date| parse(date) }
+      payment_window_dates = (calculated_start_date..(today + 6.days)).map { |date| parse(date) }
+      @chargeable_dates ||= if !@second_week_selected
+                              payment_window_dates
+                            else
+                              disable_already_chosen_dates(payment_window_dates)
+                            end
```
Comment (likeuwill) [inline app/services/dates/weekly.rb:30]:
```suggestion
      @chargeable_dates ||= begin
            payment_window_dates = (calculated_start_date..(today + 6.days)).map { |date| parse(date) }
            if !@second_week_selected
                              payment_window_dates
                            else
                              disable_already_chosen_dates(payment_window_dates)
                            end
                      end
```

File: app/services/session_manipulation/set_selected_week.rb [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+module SessionManipulation
+  ##
+  # Service used to set currently selected week for weekly payments.
+  #
+  # ==== Usage
+  #    SessionManipulation::SetSelectedWeek.call(session: session, second_week_selected: true)
+  #
+  class SetSelectedWeek < BaseManipulator
+    # Level used to clearing keys in the session
+    LEVEL = 16
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:12]: you should user a proper key, take a look into `base_manipulator`
```
      12 => %w[confirm_weekly_charge_today weekly_dates weekly_charge_today],
      13 => %w[dates total_charge weekly second_week_selected],
```
add key `second_week_selected` to array and set proper level

File: app/services/session_manipulation/set_selected_week.rb [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+      session[:second_week_dates_selected] = @second_week_selected
+
+      session[:second_week_start_date] = nil if @second_week_selected == false
+    end
+
+    private
+
+    def reset_second_week_date
+      session[:second_week_start_date] = nil
+    end
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:36]: no need

File: app/services/session_manipulation/set_selected_week.rb [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+    #
+    def initialize(session:, second_week_selected: false)
+      @session = session
+      @second_week_selected = second_week_selected
+    end
+
+    # Adds the +second_week_selected+ to the session. Used by the class level method +.call+
+    def call
+      session[:second_week_selected] = @second_week_selected
+
```
Comment (likeuwill) [inline app/services/session_manipulation/set_selected_week.rb:28]: remove empty line

File: app/services/dates/validate_selected_weekly_date.rb [hunk 25]
```diff
@@ -62,7 +65,17 @@ def date_chargeable?
...
+
+    ##
+    # Checks if date was already selected in previous week selection
+    # Returns boolean
+    def already_selected?
+      return if @session[:second_week_selected] == false
+
+      formatted_date = Time.zone.parse(@start_date).strftime(VALUE_DATE_FORMAT)
+      @session.dig(:vehicle_details, 'dates')&.include?(formatted_date)
     end
```
Comment (likeuwill) [inline app/services/dates/validate_selected_weekly_date.rb:79]: i think only `valid?` is public method if answer is yes so add here `private` block below `already_selected`. After that change you need to fix test, we can't test private methods but still u can call `valid?` and check if `already_selected?` works correctly

File: app/services/dates/validate_selected_weekly_date.rb [hunk 26]
```diff
@@ -71,9 +84,17 @@ def valid?
     def error
       if !@start_date
         I18n.t('empty', scope: 'dates.weekly')
+      elsif already_selected?
+        I18n.t('already_selected', scope: 'dates.weekly')
       else
         I18n.t('not_available', scope: 'dates.weekly')
       end
     end
+
```
Comment (likeuwill) [inline app/services/dates/validate_selected_weekly_date.rb:93]: rdoc

File: spec/requests/dates_controller/confirm_date_weekly_spec.rb [hunk 27]
```diff
@@ -17,17 +17,23 @@
   end
   let(:vrn) { 'CU123AB' }
   let(:la_id) { SecureRandom.uuid }
-  let(:params) { { 'date_year' => '2019', 'date_month' => '11', 'date_day' => '01' } }
+  let(:params) { { 'date_year' => '2019', 'date_month' => '11', 'date_day' => '01', 'week' => 'first' } }
+
+  let(:referer) { 'http://www.example.com/dates/select_weekly_date' }
```
Comment (likeuwill) [inline spec/requests/dates_controller/confirm_date_weekly_spec.rb:22]: can be removed

File: spec/requests/dates_controller/confirm_date_weekly_spec.rb [hunk 28]
```diff
@@ -17,17 +17,21 @@
   end
   let(:vrn) { 'CU123AB' }
   let(:la_id) { SecureRandom.uuid }
-  let(:params) { { 'date_year' => '2019', 'date_month' => '11', 'date_day' => '01' } }
+  let(:params) { { 'date_year' => '2019', 'date_month' => '11', 'date_day' => '01', 'week' => 'first' } }
```
Comment (likeuwill) [inline spec/requests/dates_controller/confirm_date_weekly_spec.rb:20]: test

**Pull request merged**